### PR TITLE
App 1217 - JIRA's OAuth1 implementation

### PR DESCRIPTION
**Pull request by campidelli**

### CODE REVIEW

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+ * URL. It should also read the application public key configured on the filesystem and validate it.
+ *
+ * Created by rsanchez on 24/07/17.
+ */
+@Component
+public class JiraAuthorizationManager {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(JiraAuthorizationManager.class);
+
+  private static final String COMPONENT = "JIRA Authentication Manager";
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java]: We should change the component to "JIRA Authorization Manager"

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+   * Verify if the passed user has authorized us to perform Jira API calls on behalf of him/her.
+   * @param settings Jira integration settings.
+   * @param url Jira base URL.
+   * @param userId Symphony user ID.
+   * @return <code>true</code> If the passed user has authorized the access.
+   * @throws AuthorizationException Thrown in case of error.
+   */
+  public boolean isUserAuthorized(IntegrationSettings settings, String url, Long userId)
+      throws AuthorizationException {
+    UserAuthorizationData u =
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java]: We should rename this variable.

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+    String consumerKey = (String) appAuthorizationModel.getProperties().get(CONSUMER_KEY);
+    String privateKey = getPrivateKey(settings);
+    String callbackUrl = getCallbackUrl(settings);
+
+    JiraOAuth1Provider provider = context.getBean(JiraOAuth1Provider.class);
+    provider.configure(consumerKey, privateKey, url, callbackUrl);
+
+    try {
+      URL myselfUrl = new URL(url);
+      myselfUrl = new URL(myselfUrl, "/rest/api/2/myself");
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java]: Move it to a String constant.

Something like that:
URL mySelfUrl = new URL(new URL(url), MY_SELF_PATH);

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+   */
+  public String getAuthorizationUrl(IntegrationSettings settings, String url, Long userId)
+      throws AuthorizationException {
+    AppAuthorizationModel appAuthorizationModel = getAuthorizationModel(settings);
+    String consumerKey = (String) appAuthorizationModel.getProperties().get(CONSUMER_KEY);
+    String privateKey = getPrivateKey(settings);
+    String callbackUrl = getCallbackUrl(settings);
+
+    JiraOAuth1Provider provider = context.getBean(JiraOAuth1Provider.class);
+    provider.configure(consumerKey, privateKey, url, callbackUrl);
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java]: Provider configuration code is duplicated.

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+    JiraOAuth1Data jiraOAuth1Data = new JiraOAuth1Data(temporaryToken);
+    UserAuthorizationData userAuthData = new UserAuthorizationData(url, userId, jiraOAuth1Data);
+
+    authRepoService.save(settings.getType(), settings.getConfigurationId(), userAuthData);
+
+    return authorizationUrl;
+  }
+
+  /**
+   * Authorize a temporaty token by getting a permanent access token and saving it.
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java]: temporary token

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+    }
+    UserAuthorizationData userAuthData = result.get(0);
+
+    AppAuthorizationModel appAuthorizationModel = getAuthorizationModel(settings);
+    String consumerKey = (String) appAuthorizationModel.getProperties().get(CONSUMER_KEY);
+    String privateKey = getPrivateKey(settings);
+    String callbackUrl = getCallbackUrl(settings);
+
+    JiraOAuth1Provider provider = context.getBean(JiraOAuth1Provider.class);
+    provider.configure(consumerKey, privateKey, userAuthData.getUrl(), callbackUrl);
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java]: Provider configuration code is duplicated.

File: src/main/java/org/symphonyoss/integration/jira/webhook/JiraWebHookIntegration.java [hunk 7]
```diff
@@ -93,28 +100,64 @@ public Message parse(WebHookPayload input) throws WebHookParseException {
...
 
+  /**
+   * @see AuthorizedIntegration#authorize(AuthorizationPayload)
+   */
   @Override
-  public void verifyUserAuthorizationData(UserAuthorizationData authData) {
-    Object data = authData.getData();
+  public void authorize(AuthorizationPayload authorizationPayload) throws AuthorizationException {
+    String temporaryToken = authorizationPayload.getParameters().get("oauth_token");
+    String verificationCode = authorizationPayload.getParameters().get("oauth_verifier");
```
Comment (robson-sanchez) [inline src/main/java/org/symphonyoss/integration/jira/webhook/JiraWebHookIntegration.java]: We should define as string constants.

File: src/main/resources/integration-jira-log-messages.properties [hunk 8]
```diff
@@ -2,3 +2,13 @@ integration.jira.url.base.invalid=Invalid base URL {0}.
 integration.jira.url.base.invalid.solution=Inform a valid base URL.
 integration.jira.url.authorization.invalid=Invalid authorization callback URL {0}.
 integration.jira.url.authorization.invalid.solution=Inform a valid authorization callback URL.
+integration.jira.public.key.validation=Unexpected error when validate the application public key.
+integration.jira.private.key.validation=Unexpected error when validate the application private key.
+integration.jira.url.api.invalid=Invalid API URL {0}.
+integration.jira.url.api.invalid.solution=Inform a valid API URL.
+integration.jira.auth.user.data.not.found=No auth user data found for the informed temporary token {0}.
+integration.jira.auth.user.data.not.found.solution=Start the authentication flow by getting the authentication URL.
```
Comment (robson-sanchez) [inline src/main/resources/integration-jira-log-messages.properties]: Start the authentication flow by getting the authorization URL.

File: src/main/resources/integration-jira-log-messages.properties [hunk 9]
```diff
@@ -2,3 +2,13 @@ integration.jira.url.base.invalid=Invalid base URL {0}.
...
+integration.jira.authorize.insufficient.params=Insufficient parameters.
+integration.jira.authorize.insufficient.params.solution=Both parameters 'oauth_token' and 'oauth_verifier' must be informed.
+integration.jira.authorize.no.integration.settings=Integration settings not found.
+integration.jira.authorize.no.integration.settings.solution=Check the parameters used to load the integration settings file.
```
Comment (robson-sanchez) [inline src/main/resources/integration-jira-log-messages.properties]: We should change to: 'Make sure the integration was bootstrapped successfully'

File: src/main/resources/integration-jira-log-messages.properties [hunk 10]
```diff
@@ -2,3 +2,13 @@ integration.jira.url.base.invalid=Invalid base URL {0}.
 integration.jira.url.base.invalid.solution=Inform a valid base URL.
 integration.jira.url.authorization.invalid=Invalid authorization callback URL {0}.
 integration.jira.url.authorization.invalid.solution=Inform a valid authorization callback URL.
+integration.jira.public.key.validation=Unexpected error when validate the application public key.
+integration.jira.private.key.validation=Unexpected error when validate the application private key.
+integration.jira.url.api.invalid=Invalid API URL {0}.
```
Comment (robson-sanchez) [inline src/main/resources/integration-jira-log-messages.properties]: 'Invalid JIRA Base URL {0}'

File: src/main/resources/integration-jira-log-messages.properties [hunk 11]
```diff
@@ -2,3 +2,13 @@ integration.jira.url.base.invalid=Invalid base URL {0}.
...
+integration.jira.url.api.invalid.solution=Inform a valid API URL.
```
Comment (robson-sanchez) [inline src/main/resources/integration-jira-log-messages.properties]: 'Make sure the MessageML is describing the base URL correctly.'

**Review by robson-sanchez — CHANGES_REQUESTED** [review]
Codecov must report 90% code coverage

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java (hunk 1 above)
Comment (campidelli) [reply src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java]: Done.

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java (hunk 2 above)
Comment (campidelli) [reply src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java]: Done.

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java (hunk 3 above)
Comment (campidelli) [reply src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java]: Done.

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java (hunk 4 above)
Comment (campidelli) [reply src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java]: Done.

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java (hunk 5 above)
Comment (campidelli) [reply src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java]: Done.

File: src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java (hunk 6 above)
Comment (campidelli) [reply src/main/java/org/symphonyoss/integration/jira/authorization/JiraAuthorizationManager.java]: Done.

File: src/main/java/org/symphonyoss/integration/jira/webhook/JiraWebHookIntegration.java (hunk 7 above)
Comment (campidelli) [reply src/main/java/org/symphonyoss/integration/jira/webhook/JiraWebHookIntegration.java]: Done.

File: src/main/resources/integration-jira-log-messages.properties (hunk 10 above)
Comment (campidelli) [reply src/main/resources/integration-jira-log-messages.properties]: Done.

File: src/main/resources/integration-jira-log-messages.properties (hunk 11 above)
Comment (campidelli) [reply src/main/resources/integration-jira-log-messages.properties]: Done.

File: src/main/resources/integration-jira-log-messages.properties (hunk 8 above)
Comment (campidelli) [reply src/main/resources/integration-jira-log-messages.properties]: Done.

File: src/main/resources/integration-jira-log-messages.properties (hunk 9 above)
Comment (campidelli) [reply src/main/resources/integration-jira-log-messages.properties]: Done.

**Review by robson-sanchez — APPROVED** [review]

**Pull request merged**